conmon/conmon.c

```diff
@@ -1339,6 +1368,9 @@ int main(int argc, char *argv[])
 		exit(127);
 	}
 
+	if (opt_exit_command)
+		atexit(do_exit_command);
```

Comment:
I still don't think we need `atexit`.  Do we want this firing if `conmon` dies during some later error?  I'd rather call `do_exit_command` directly where `main` currently returns `EXIT_SUCCESS`.

Reply:
If the intent is to perform general cleanup, firing the command even after errors is useful if we can guarantee the container is dead (firing after an internal conmon error when the container is still alive would be undesirable). We could afford to move the `atexit` a little lower, where we're certain the container has started successfully - anything before that point we can catch easily from whatever launched `conmon`.

Reply:
I am not sure what should happen to the container when conmon fails internally, should the container still be considered running if we cannot access it?

I think the safest thing to do is to let the container process die as well if conmon fails (`PR_SET_PDEATHSIG`)

@mrunalp @rhatdan what do you think?

Reply:
Until [here][1], there's no exit-dir file.  That is very close to [where I think we want to call `do_exit_command][2], no need for `atexit`.

`PR_SET_PDEATHSIG` sounds reasonable to me too.

[1]: https://github.com/kubernetes-incubator/cri-o/blob/8ea79e755f5f89b132e642015a46d9d291b246ef/conmon/conmon.c#L1472
[2]: https://github.com/kubernetes-incubator/cri-o/blob/8ea79e755f5f89b132e642015a46d9d291b246ef/conmon/conmon.c#L1485

Reply:
if we start using `PR_SET_PDEATHSIG` then I fully agree we should drop the `atexit` usage.

Otherwise I am not sure we should not notify the cleanup process that conmon is exiting and there won't be any other notification (the hook can still check if the exit file exists and act accordingly)

Reply:
I think `PR_SET_PDEATHSIG` is a good idea, considering we have no way of monitoring the container if `conmon` dies.

Reply:
unfortunately `PR_SET_PDEATHSIG` won't work here as it is reset on a `fork(2)`, so the process won't get the signal once runc forks.

At this point I wonder, should we try to catch early exits, if conmon fails for any reason, an do a final `$RUNTIME delete -f $CONTAINER_UUID`?

Reply:
I am still convinced that the cleanup process should be called even if the exit file doesn't exist.  It is the hook responsibility to check that the file exists and it won't miss the notification that conmon is terminating.

Reply:
> I am still convinced that the cleanup process should be called even if the exit file doesn't exist.

I still don't think we need `atexit` for that.  Can't we `goto cleanup` or some such with a cleanup block at the end of `main`?  Or is there nothing else to cleanup?

Reply:
> At this point I wonder, should we try to catch early exits, if conmon fails for any reason, an do a final `kill -9 container_pid`?

With early errors also leading somehow to exit-cmd invocation, this sort of meta cleanup could happen there.

Reply:
we will need to call the cleanup function from every place where we `exit(3)`: such as `pexit` and `nexit` and we will end up duplicating the functionality of `atexit(3)` that is part of the C library.  Why not to use it?

Reply:
> we will need to call the cleanup function from every place where we `exit(3)`: such as `pexit` and `nexit`...

We are currently ver liberal with those.  I'd be happier if places like [this][1] logged an error and returned an error to their caller.  If we keep exiting there, we'll want `atexit` handlers for all cleanup, including [this][2].  I think that cleanup is easier with `goto cleanup`, because you don't need helper functions, but `atexit` would work if maintainers prefer it for some reason.

[1]: https://github.com/kubernetes-incubator/cri-o/blob/master/conmon/conmon.c#L1066
[2]: https://github.com/kubernetes-incubator/cri-o/blob/master/conmon/conmon.c#L1481